In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### The 5 parameters for the last race and the 5 parameters for the race before that and 5 parameters for race previous to that (i.e. last, second_last, third_last) for each horse

In [193]:
runs_df = pd.read_csv("runs.csv")
races_df = pd.read_csv("races.csv")

In [194]:
runs_df.sample(5)

race_id  horse_no  horse_id  result  won  lengths_behind  horse_age  \
30837     2464         3      2572       8  0.0            3.00          3   
32269     2576        10      2160      11  0.0            5.25          3   
12186      972        10      1081       9  0.0            6.00          3   
18866     1505         7      2361       8  0.0            4.50          3   
44859     3581         9      2765       8  0.0            5.25          3   

      horse_country horse_type  horse_rating  ...  time2  time3  time4  time5  \
30837           AUS    Gelding            60  ...  24.52  25.18  23.90    NaN   
32269           IRE    Gelding            60  ...  23.34  24.04  25.17    NaN   
12186           IRE    Gelding            60  ...  22.62  24.07  24.83  25.06   
18866            GB      Horse            60  ...  23.31  22.56    NaN    NaN   
44859           AUS    Gelding            60  ...  21.28  22.87    NaN    NaN   

       time6  finish_time  win_odds  place_odds  trainer_id  jockey_id  
30837    NaN       102.04      27.0         7.3         164        135  
32269    NaN       100.65       6.1         2.1          29        132  
12186    NaN       110.80      19.0         5.4          80         64  
18866    NaN        70.65      71.0        16.0          69         34  
44859    NaN        58.01      99.0        21.0          55         18  

[5 rows x 37 columns]

In [195]:
races_df.sample(5)

race_id        date venue  race_no config  surface  distance  \
1176     1176  1998-12-02    HV        4      B        0      1200   
2686     2686  2000-11-19    ST        7    B+2        0      1400   
3511     3511  2001-12-17    ST        9    B+2        0      1200   
6207     6207  2005-07-06    HV        5      C        0      1650   
5930     5930  2005-01-19    HV        6    C+3        0      1000   

                 going horse_ratings      prize  ...  place_combination3  \
1176              GOOD         60-40   625000.0  ...                 8.0   
2686              GOOD             G  6000000.0  ...                 6.0   
3511              GOOD         80-60   950000.0  ...                12.0   
6207  GOOD TO YIELDING         80-60  1235000.0  ...                 3.0   
5930              GOOD         80-60  1165000.0  ...                 8.0   

      place_combination4  place_dividend1  place_dividend2  place_dividend3  \
1176                 NaN             21.5             16.5             27.0   
2686                 NaN             11.5             12.0             26.5   
3511                 NaN             27.5             15.0             13.5   
6207                 NaN             26.0             15.0             68.5   
5930                 NaN             24.0             19.5             15.0   

      place_dividend4  win_combination1  win_dividend1  win_combination2  \
1176              NaN                10           82.5               NaN   
2686              NaN                 2           22.5               NaN   
3511              NaN                 5           91.0               NaN   
6207              NaN                 4           97.0               NaN   
5930              NaN                12           67.5               NaN   

      win_dividend2  
1176            NaN  
2686            NaN  
3511            NaN  
6207            NaN  
5930            NaN  

[5 rows x 37 columns]

In [211]:
horse = pd.merge(runs_df, races_df,on='race_id')

In [212]:
runs_df.shape

(79447, 37)

In [213]:
races_df.shape

(6349, 37)

In [214]:
horse.shape

(79447, 73)

In [215]:
# Number of days between current race and last race(days_lastrace)

horse.date = pd.to_datetime(horse.date)

In [216]:
horse['date'].head()

0   1997-06-02
1   1997-06-02
2   1997-06-02
3   1997-06-02
4   1997-06-02
Name: date, dtype: datetime64[ns]

In [217]:
horse['date'].tail()

79442   2005-08-28
79443   2005-08-28
79444   2005-08-28
79445   2005-08-28
79446   2005-08-28
Name: date, dtype: datetime64[ns]

In [218]:
unique_horses = horse.horse_id.unique()
unique_horses = list(unique_horses)

In [219]:
last_race = pd.DataFrame()
second_last_race = pd.DataFrame()
third_last_race = pd.DataFrame()

# For the Last race
+ created new dataframe named 'last_race' with race details for all horses

In [273]:
horse[horse['horse_id'] == 3917].iloc[-1:,:]['date'] - horse[horse['horse_id'] == 3917].iloc[-2:-1,:]['date']

19168   NaT
20419   NaT
Name: date, dtype: timedelta64[ns]

###### 

In [271]:
horse[horse['horse_id'] == 3917].iloc[-1:,:]['date'].dt - horse[horse['horse_id'] == 3917].iloc[-2:-1,:]['date'].dt

TypeError: unsupported operand type(s) for -: 'DatetimeProperties' and 'DatetimeProperties'

In [253]:
d2 = horse[horse['horse_id'] == 3917].iloc[-2:-1,:]['date']

In [255]:
(horse[horse['horse_id'] == 3917].iloc[-1:,:]['date'] -  horse[horse['horse_id'] == 3917].iloc[-2:-1,:]['date']).days

AttributeError: 'Series' object has no attribute 'days'

In [241]:
horse[horse['horse_id'] == 3917].iloc[-1:,:]['date'] - horse[horse['horse_id'] == 3917].iloc[-2:-1,:]['date']

19168   NaT
20419   NaT
Name: date, dtype: timedelta64[ns]

In [220]:
for i in unique_horses:
    last_race = last_race.append(horse[horse['horse_id'] == i].iloc[-1:,:])

In [227]:
last_race[last_race['horse_id'] == 3917]['date']

20419   1999-07-28
Name: date, dtype: datetime64[ns]

# For the second Last race
+ created new dataframe named 'second_last_race' with race details for all horses

In [133]:
for i in unique_horses:
    second_last_race = second_last_race.append(horse[horse['horse_id'] == i].iloc[-2:-1,:])

# For the third Last race
+ created new dataframe named 'third_last_race' with race details for all horses

In [134]:
for i in unique_horses:
    third_last_race = third_last_race.append(horse[horse['horse_id'] == i].iloc[-3:-2,:])

In [135]:
last_race.shape

(4405, 73)

In [136]:
second_last_race.shape

(4023, 73)

*From the above observation we can say that there 382 horse who have participated in only one race.*

In [137]:
third_last_race.shape

(3823, 73)

### Adding required 5 parameter for last race for each horse

In [187]:
horse.date

0       1997-06-02
1       1997-06-02
2       1997-06-02
3       1997-06-02
4       1997-06-02
5       1997-06-02
6       1997-06-02
7       1997-06-02
8       1997-06-02
9       1997-06-02
10      1997-06-02
11      1997-06-02
12      1997-06-02
13      1997-06-02
14      1997-06-02
15      1997-06-02
16      1997-06-02
17      1997-06-02
18      1997-06-02
19      1997-06-02
20      1997-06-02
21      1997-06-02
22      1997-06-02
23      1997-06-02
24      1997-06-02
25      1997-06-02
26      1997-06-02
27      1997-06-02
28      1997-06-02
29      1997-06-02
           ...    
79417   2005-08-28
79418   2005-08-28
79419   2005-08-28
79420   2005-08-28
79421   2005-08-28
79422   2005-08-28
79423   2005-08-28
79424   2005-08-28
79425   2005-08-28
79426   2005-08-28
79427   2005-08-28
79428   2005-08-28
79429   2005-08-28
79430   2005-08-28
79431   2005-08-28
79432   2005-08-28
79433   2005-08-28
79434   2005-08-28
79435   2005-08-28
79436   2005-08-28
79437   2005-08-28
79438   2005

In [157]:
final_df = pd.DataFrame()

In [158]:
final_df[['race_id','horse_id','horse_age','date','finish_position','class','condition','length_behind','finish_time']] = last_race[['race_id',\
                 'horse_id','horse_age','date','result','race_class','config','lengths_behind','finish_time']]

In [159]:
final_df.head()

race_id  horse_id  horse_age       date  finish_position  class  \
20419     1629      3917          3 1999-07-28               12      5   
21880     1745      2157          3 1999-09-13                8      5   
14961     1191       858          3 1998-12-09                9      5   
3709       295      1853          3 1997-09-27               14      5   
9895       791      2796          3 1998-06-24                3      5   

      condition  length_behind  finish_time  
20419         B          14.75       140.21  
21880       A+3          11.50       111.83  
14961         A          12.50        72.04  
3709          A          12.25       112.29  
9895          A           2.25       101.70

In [160]:
final_df.shape

(4405, 9)

### Adding required 5 parameter for second last race for each horse

In [161]:
final_df1 = pd.DataFrame()

In [163]:
final_df1[['race_id1','horse_id','date1','finish_position1','class1','condition1','length_behind1','finish_time1']] = second_last_race[\
            ['race_id','horse_id','date','result','race_class','config','lengths_behind','finish_time']]

In [164]:
final_df1.head()

race_id1  horse_id      date1  finish_position1  class1 condition1  \
19168      1530      3917 1999-06-13                 7       5        B+2   
21550      1719      2157 1999-09-01                 9       5          C   
13708      1091       858 1998-10-25                11       4          C   
3170        253      1853 1997-09-14                11       5        A+3   
9200        735      2796 1998-06-03                 2       5          A   

       length_behind1  finish_time1  
19168            6.25        109.60  
21550            5.75        103.96  
13708            5.25         86.39  
3170             5.50         71.48  
9200             0.25        100.84

In [165]:
final_df1.shape

(4023, 8)

### Adding required 5 parameter for third last race for each horse

In [166]:
final_df2 = pd.DataFrame()

In [168]:
final_df2[['race_id2','horse_id','date2','finish_position2','class2','condition2','length_behind2','finish_time2']] = third_last_race[\
                        ['race_id','horse_id','date','result','race_class','config','lengths_behind','finish_time']]

In [169]:
final_df2.head()

race_id2  horse_id      date2  finish_position2  class2 condition2  \
18728      1492      3917 1999-05-30                 8       5        C+3   
20414      1629      2157 1999-07-28                11       5          B   
12543       999       858 1998-09-18                 4       4        B+2   
2140        171      1853 1997-08-10                 6       5          A   
8318        667      2796 1998-02-25                 2       5        C+3   

       length_behind2  finish_time2  
18728            7.50         97.89  
20414           14.50        140.20  
12543            2.50         70.54  
2140             1.75        101.06  
8318             0.50        101.97

In [170]:
final_df2.shape

(3823, 8)

In [171]:
new_df = final_df.merge(final_df1,how='left',on='horse_id')
#pd.merge(pd.merge(df1,df2,on='name'),df3,on='name')

In [173]:
new_df.head()

race_id  horse_id  horse_age       date  finish_position  class condition  \
0     1629      3917          3 1999-07-28               12      5         B   
1     1745      2157          3 1999-09-13                8      5       A+3   
2     1191       858          3 1998-12-09                9      5         A   
3      295      1853          3 1997-09-27               14      5         A   
4      791      2796          3 1998-06-24                3      5         A   

   length_behind  finish_time  race_id1      date1  finish_position1  class1  \
0          14.75       140.21    1530.0 1999-06-13               7.0     5.0   
1          11.50       111.83    1719.0 1999-09-01               9.0     5.0   
2          12.50        72.04    1091.0 1998-10-25              11.0     4.0   
3          12.25       112.29     253.0 1997-09-14              11.0     5.0   
4           2.25       101.70     735.0 1998-06-03               2.0     5.0   

  condition1  length_behind1  finish_time1  
0        B+2            6.25        109.60  
1          C            5.75        103.96  
2          C            5.25         86.39  
3        A+3            5.50         71.48  
4          A            0.25        100.84

In [174]:
new_df.shape

(4405, 16)

In [175]:
horse_df = new_df.merge(final_df2,how='left',on='horse_id')

In [177]:
horse_df.head()

race_id  horse_id  horse_age       date  finish_position  class condition  \
0     1629      3917          3 1999-07-28               12      5         B   
1     1745      2157          3 1999-09-13                8      5       A+3   
2     1191       858          3 1998-12-09                9      5         A   
3      295      1853          3 1997-09-27               14      5         A   
4      791      2796          3 1998-06-24                3      5         A   

   length_behind  finish_time  race_id1  ... condition1  length_behind1  \
0          14.75       140.21    1530.0  ...        B+2            6.25   
1          11.50       111.83    1719.0  ...          C            5.75   
2          12.50        72.04    1091.0  ...          C            5.25   
3          12.25       112.29     253.0  ...        A+3            5.50   
4           2.25       101.70     735.0  ...          A            0.25   

   finish_time1 race_id2      date2  finish_position2  class2 condition2  \
0        109.60   1492.0 1999-05-30               8.0     5.0        C+3   
1        103.96   1629.0 1999-07-28              11.0     5.0          B   
2         86.39    999.0 1998-09-18               4.0     4.0        B+2   
3         71.48    171.0 1997-08-10               6.0     5.0          A   
4        100.84    667.0 1998-02-25               2.0     5.0        C+3   

   length_behind2  finish_time2  
0            7.50         97.89  
1           14.50        140.20  
2            2.50         70.54  
3            1.75        101.06  
4            0.50        101.97  

[5 rows x 23 columns]

In [178]:
horse_df.shape

(4405, 23)

# In our merged dataset for last race, second last race and race previous to that has 'NaN' values at some places because there are some horses participated in only one or two race.

* `We will replace those 'NaN' values to '-' for further analysis.`

In [179]:
ndf = horse_df.replace(np.nan,'-')

In [180]:
ndf.tail()

race_id  horse_id  horse_age       date  finish_position  class  \
4400     6345      2607          3 2005-08-28               10     11   
4401     6346      2605          3 2005-08-28                7     11   
4402     6346      2606          3 2005-08-28                9     11   
4403     6347      2093          3 2005-08-28               10      2   
4404     6348       216          3 2005-08-28                1      2   

     condition  length_behind  finish_time race_id1  ... condition1  \
4400         A           3.50        94.02        -  ...          -   
4401         A           4.75       121.73        -  ...          -   
4402         A           7.00       122.08        -  ...          -   
4403         A           6.25        70.14        -  ...          -   
4404         A           0.00        95.96        -  ...          -   

     length_behind1 finish_time1 race_id2 date2 finish_position2 class2  \
4400              -            -        -     -                -      -   
4401              -            -        -     -                -      -   
4402              -            -        -     -                -      -   
4403              -            -        -     -                -      -   
4404              -            -        -     -                -      -   

     condition2 length_behind2 finish_time2  
4400          -              -            -  
4401          -              -            -  
4402          -              -            -  
4403          -              -            -  
4404          -              -            -  

[5 rows x 23 columns]

In [182]:
horse_df.condition.value_counts()

A      1766
C       728
C+3     664
B       535
B+2     370
A+3     342
Name: condition, dtype: int64

In [183]:
horse_df['class'].value_counts()

4     1479
5     1037
3     1019
2      311
11     301
1      140
13      78
12      22
6       15
0        3
Name: class, dtype: int64

## Let's check for horse which participated in current 3 races

+ We will drop horses from our dataset which played only one or two race

In [184]:
ndf = ndf.loc[~ndf.apply(lambda row: (row=='-').any(), axis=1)]

In [185]:
ndf.shape

(3823, 23)

In [186]:
ndf.tail()

race_id  horse_id  horse_age       date  finish_position  class  \
4320     6304      2595          3 2005-08-14                2      3   
4321     6339       458          3 2005-08-28               14      3   
4323     6341      1543          3 2005-08-28                2      3   
4324     6341      3331          3 2005-08-28                1      3   
4343     6341        87          3 2005-08-28               14      3   

     condition  length_behind  finish_time race_id1  ... condition1  \
4320         C            1.0        69.61     6287  ...        B+2   
4321         A            8.5        70.78     6300  ...        C+3   
4323         A            0.1        82.17     6310  ...          C   
4324         A            0.0        82.16     6287  ...        B+2   
4343         A            8.0        83.43     6318  ...          A   

     length_behind1 finish_time1 race_id2                date2  \
4320           6.25        82.76     6249  1122163200000000000   
4321           8.75        71.93     6249  1122163200000000000   
4323           0.75         81.9     6252  1122163200000000000   
4324           1.25        81.95     6254  1122163200000000000   
4343           15.5       112.19     6274  1122681600000000000   

     finish_position2 class2 condition2 length_behind2 finish_time2  
4320               11      3        C+3           5.25        69.97  
4321                4      3        C+3           1.75        69.44  
4323                1      4        C+3              0        82.75  
4324                1      3        C+3              0        82.24  
4343               14      3        A+3              9        96.46  

[5 rows x 23 columns]

In [37]:
ndf['class'].value_counts()

4     1302
5     1025
3      920
2      292
1      132
11      79
13      43
12      21
6        7
0        2
Name: class, dtype: int64

In [38]:
ndf.condition.value_counts()

A      1472
C       653
C+3     574
B       500
B+2     325
A+3     299
Name: condition, dtype: int64

## Below are the details for each horse participated in race with `race class 4 and condition 'A'`

In [39]:
a = ndf.groupby('class').get_group(4).groupby('condition').get_group('A')[['horse_id','finish_position','finish_position1','finish_position2']]

In [40]:
a.shape

(442, 4)

In [41]:
k = []
for i, j in a.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

152

In [42]:
l = []
for i, j in a.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

39

# from the above observation we can say that,
- There are 442 horse which participated all last 3 races with race class = 4 and condition = 'A'.
- Out of which only 39 horses has shown the improvement in final position i.e. they have better achieve better final position as compared to previous race.
- Also if we compare only for last two race then there are 152 horses which has imporvement in their final position

In [43]:
print((39/442)*100)

8.823529411764707


### Below is the record of horses which shown improvement in final postion in last 3 races

In [44]:
a[a['horse_id'].isin(l)]

horse_id  finish_position finish_position1 finish_position2
70        3859                8                9               12
116       3196                7                9               11
154       1581                8               11               14
205       2995                4                8                9
257       1677                5                8               11
488       2962                3                7                9
566       3620                1                7                8
1359      2957                9               10               11
1618      3318                1                2                4
1639      1452               12               13               14
1729      1878                2                4                7
1766      2885                2                3                9
1887      1593               11               13               14
1949      3223                3                5               13
2112       760                5                9               13
2297      2226                1                9               10
2408      2284                6                9               10
2449      4290                5                6               11
2663      4078               10               11               12
3061      1621                3                6                8
3084      4024                1                5                6
3118      1233                1                2                4
3146      2712                2                6                7
3290      3796                1                4                6
3333      1317                9               11               14
3419      2529                4                6                7
3763      2652                7               10               11
3862      1394                2               11               12
3865      4383                7                8               11
3879       938                4                6               13
3905      2451                5                8               10
3942      1838                1                7               10
3964      1214                4                8               12
3971      3461                9               12               13
4109       528                1                7               14
4122      1998                9               10               13
4165      2370                7                8               14
4283      4360                3                5               11
4287      1956                5               10               11

## Below are the details for each horse participated in race with `race class 4 and condition 'C'`

In [45]:
a = ndf.groupby('class').get_group(4).groupby('condition').get_group('C')[['horse_id','finish_position','finish_position1','finish_position2']]

In [46]:
a.shape

(239, 4)

In [47]:
k = []
for i, j in a.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

86

In [48]:
l = []
for i, j in a.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

29

# from the above observation we can say that,
- There are 239 horse which participated all last 3 races with race class = 4 and condition = 'C'.
- Out of which only 29 horses has shown the improvement in final position i.e. they have better achieve better final position as compared to previous race
- Also if we compare only for last two race then there are 86 horses which has imporvement in their final position

In [49]:
print((29/239)*100)

12.133891213389122


## Below are the details for each horse participated in race with `race class 5 and condition 'A'`

In [50]:
a = ndf.groupby('class').get_group(5).groupby('condition').get_group('A')[['horse_id','finish_position','finish_position1','finish_position2']]

In [51]:
a.shape

(402, 4)

In [52]:
k = []
for i, j in a.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

155

In [53]:
l = []
for i, j in a.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

41

# from the above observation we can say that,
- There are 402 horse which participated all last 3 races with race class = 4 and condition = 'A'.
- Out of which only 41 horses has shown the improvement in final position i.e. they have better achieve better final position as compared to previous race
- Also if we compare only for last two race then there are 155 horses which has imporvement in their final position

In [54]:
print((41/402)*100)

10.199004975124378


## Below are the details for each horse participated in race with `race class 5 and condition 'C'`

In [55]:
a = ndf.groupby('class').get_group(5).groupby('condition').get_group('C')[['horse_id','finish_position','finish_position1','finish_position2']]

In [56]:
a.shape

(155, 4)

In [57]:
k = []
for i, j in a.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

58

In [58]:
l = []
for i, j in a.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

20

# from the above observation we can say that,
- There are 155 horse which participated all last 3 races with race class = 5 and condition = 'C'.
- Out of which only 20 horses has shown the improvement in final position i.e. they have better achieve better final position as compared to previous race
- Also if we compare only for last two race then there are 58 horses which has imporvement in their final position

In [59]:
print((20/155)*100)

12.903225806451612


+ <font color=red>We have compared data based on major class and condition value and obeserve that around 8-12 % horses has shown improvement with respect to final position.</font>

# <font color= Red>Same result we can see if compared all horses irrespective of race class and condition of race</font>

In [60]:
ndf.shape

(3823, 19)

In [61]:
k = []
for i, j in ndf.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

1392

In [62]:
l = []
for i, j in ndf.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

350

# Obeservation for entire dataset,
- There are 2823 horse which participated all last 3 races.
- Out of which 350 horses has shown the improvement in final position i.e. they have better achieve better final position as compared to previous race
- Also if we compare only for last two race then there are 1392 horses which has imporvement in their final position

In [63]:
print((350/3823)*100)

9.155113784985613


In [64]:
print((1392/3823)*100)

36.41119539628564


In [65]:
last_race.columns

Index(['race_id', 'horse_no', 'horse_id', 'result', 'won', 'lengths_behind',
       'horse_age', 'horse_country', 'horse_type', 'horse_rating',
       'horse_gear', 'declared_weight', 'actual_weight', 'draw',
       'position_sec1', 'position_sec2', 'position_sec3', 'position_sec4',
       'position_sec5', 'position_sec6', 'behind_sec1', 'behind_sec2',
       'behind_sec3', 'behind_sec4', 'behind_sec5', 'behind_sec6', 'time1_x',
       'time2_x', 'time3_x', 'time4_x', 'time5_x', 'time6_x', 'finish_time',
       'win_odds', 'place_odds', 'trainer_id', 'jockey_id', 'date', 'venue',
       'race_no', 'config', 'surface', 'distance', 'going', 'horse_ratings',
       'prize', 'race_class', 'sec_time1', 'sec_time2', 'sec_time3',
       'sec_time4', 'sec_time5', 'sec_time6', 'sec_time7', 'time1_y',
       'time2_y', 'time3_y', 'time4_y', 'time5_y', 'time6_y', 'time7',
       'place_combination1', 'place_combination2', 'place_combination3',
       'place_combination4', 'place_dividend1', 'plac

## Below are the details for each horse participated in race with same race class

In [66]:
a = ndf[(ndf['class'] == ndf['class1']) & (ndf['class'] == ndf['class2'])][['horse_id','finish_position','finish_position1','finish_position2']]

In [67]:
a.shape

(2818, 4)

In [68]:
a.head()

horse_id  finish_position finish_position1 finish_position2
0      3917               12                7                8
1      2157                8                9               11
3      1853               14               11                6
4      2796                3                2                2
6       911               11               10               10

In [69]:
k = []
for i, j in a.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

1042

In [70]:
l = []
for i, j in a.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

272

# from the above observation we can say that,
- There are `2818` horse which participated all last 3 races with race class.
- Out of which only `272` horses has shown the improvement in final position i.e. they have achieve better final position as compared to previous race
- Also if we compare only for last two race then there are `1042` horses which has imporvement in their final position

In [71]:
print((272/2818)*100)

9.65223562810504


## Below are the details for each horse participated in race with condition.

In [72]:
a = ndf[(ndf['condition'] == ndf['condition1']) & (ndf['condition1'] == ndf['condition2'])][['horse_id','finish_position','finish_position1','finish_position2']]

In [73]:
a.shape

(305, 4)

In [74]:
a.head()

horse_id  finish_position finish_position1 finish_position2
12       727               12                9               11
20      1926                2               11                4
60      3301               14               12                6
64      3105                9               13                8
90      3440                8                9                7

In [75]:
k = []
for i, j in a.iterrows():
    if (j['finish_position1'] > j['finish_position']):
        k.append(j['horse_id'])
len(k)

104

In [76]:
l = []
for i, j in a.iterrows():
    if (j['finish_position2'] > j['finish_position1'] > j['finish_position']):
        l.append(j['horse_id'])
len(l)

23

# from the above observation we can say that,
- There are `305` horse which participated all last 3 races with same condition.
- Out of which only `23` horses has shown the improvement in final position i.e. they have achieve better final position as compared to previous race
- Also if we compare only for last two race then there are `104` horses which has imporvement in their final position

In [77]:
print((23/305)*100)

7.540983606557377
